### PR TITLE
VICRegLoss Implementation

### PR DESCRIPTION
**Pull request by codeandproduce**
Solves #372.

Based on [VICReg: Variance-Invariance-Covariance Regularization for Self-Supervised Learning.
](https://arxiv.org/abs/2105.04906)

In relation to [this suggestion](https://github.com/KevinMusgrave/pytorch-metric-learning/issues/372#issuecomment-995111700), if I'm going to end up overriding both the `forward()` and `compute_loss()` function, I thought I might as well inherit the `BaseMetricLossFunction` class.

Additionally, I didn't really see the need to create separate modules for the loss components. They're just instance methods of the VICRegLoss class.

Could you take a look at the test script for this loss function though? I'm not sure if it really serves a purpose since it's pretty much identical to the actual loss function's implementation. I did write a different way to calculate the covariance loss though.

### CODE REVIEW

**Comment by codeandproduce** [comment]
@KevinMusgrave 

Hi Kevin, I was wondering if you had a chance to review this PR?

**Comment by KevinMusgrave** [comment]
I haven't gotten to it yet. I'll try to merge this and then publish v1.1 around new years day.
